node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -513,15 +544,34 @@ where
 	let rotation_info = get_group_rotation_info(ctx, relay_parent).await?;
 
 	let current_group_index = rotation_info.group_for_core(core_index, cores);
-	let current_validators = groups
-		.get(current_group_index.0 as usize)
-		.map(|v| v.as_slice())
-		.unwrap_or_default();
+
+	let preconnect_group_index = match assignment_type {
+		AssignmentType::Current => None,
+		AssignmentType::Lookahead => {
+			let next_rotation_at = rotation_info.next_rotation_at();
+			if next_rotation_at.saturating_sub(rotation_info.now) <= NEXT_GROUP_PRECONNECT_WINDOW {
+				let next_rotation_info = rotation_info.bump_rotation();
+				// Returns wrong validators group on session boundaries.
+				// In this case the peer set will be updated on `DistributeCollation`.
+				Some(next_rotation_info.group_for_core(core_index, cores))
+					.filter(|idx| *idx != current_group_index)
+			} else {
+				None
+			}
+		},
+	};
+
+	let current_validators = preconnect_group_index
```

Comment:
I don't get it. On preconnect, we connect both to the current and next group, when close (`NEXT_GROUP_PRECONNECT_WINDOW`) - so this would work even for a collation on the same block, but we deliberately make it not work on free cores - why?

Reply:
Why do we need a preconnect to work on free cores?

Reply:
So we get your `connect` and the original `pre-connect` with just one call and parachains can use it however they can.

Reply:
The thing is, why doesn't it work. From looking at the changes, it looks to me like we are making it not work deliberately and with extra effort. I would understand if we don't support it, because it makes the code more complex, but the opposite seems to be true.

Reply:
I thought we will not implement simple `Connect` on the same block, `is_collating_on_child` sounds misleading then, I'll need to reconsider this

Reply:
> The thing is, why doesn't it work. From looking at the changes, it looks to me like we are making it not work deliberately and with extra effort. I would understand if we don't support it, because it makes the code more complex, but the opposite seems to be true.

It's not hard to allow connecting on scheduled cores, but what kind of API do we provide to cumulus then? And how do we use it?

Reply:
I was actually arguing, that `Connect` and `PreConnect` can be the same thing. The API would be exactly the same, it would just be legal/accepted to return true for `is_collating_on_child` even on a free core.

Usage:
1. If the parachain knows it will provide a collation one block in advance already - return true one block before.
2. If it only knows it on the very block - return true on the block it is going to provide a collation - this still gives us up to 0.5 seconds advantage.
3. If the parachain cannot predict a collation at all, it just always returns false, resulting in the same behavior as we have now.

Reply:
And yes, the name would be misleading then.

Reply:
I see, makes sense, good thing this is simple to fix